[Bug]: `npx storybook@next init` installs the TypeScript template into a JavaScript project initialized by CRA

### Describe the bug

1. Initialize a new React project
```
npx create-react-app my-app
```

2. Install Storybook using storybook@next
```
cd my-app
npx storybook@next init
```

3. Soon after Storybook runs on the 6006 port, the following error message will appear in the terminal (part of the error message is removed for brevity):
```
WARN Force closed preview build
ModuleNotFoundError: Module not found: Error: Can't resolve './Button' in '/root/tmp/my-app/src/stories'

...

ModuleNotFoundError: Module not found: Error: Can't resolve './Header' in '/root/tmp/my-app/src/stories'

...

ModuleNotFoundError: Module not found: Error: Can't resolve './Page' in '/root/tmp/my-app/src/stories'

...
```

4. List the files under the `src` folder and files added by the `npx storybook@next init` command:
```
root@LAPTOP-7C2UU06T:~/tmp/my-app# ls src
App.css  App.test.js  index.js  reportWebVitals.js  stories
App.js   index.css    logo.svg  setupTests.js
root@LAPTOP-7C2UU06T:~/tmp/my-app# ls src/stories/
Button.stories.ts  Configure.mdx      Header.tsx       Page.tsx  button.css  page.css
Button.tsx         Header.stories.ts  Page.stories.ts  assets    header.css
```

It looks like Storybook CLI thinks the React project is a TypeScript project, so it copies TypeScript story files to the `src/stories` folder. This might be related to the new [implementation](https://github.com/storybookjs/storybook/blob/next/code/lib/cli/src/detect.ts#L161) of the `detectLanguage` function. The function checks whether `node_modules/typescript/package.json` exists and retrieves the version of TypeScript from that file if it exists. The retrieved version will then be treated by Storybook CLI as an evidence that the project is a TypeScript based project. However, just because there exists `node_modules/typescript` doesn't mean the project itself is a TypeScript project.

`npx storybook@latest init` works fine and will add JavaScript story files to a newly created CRA project.

### To Reproduce

The issue is related to Storybook CLI. The [documentation](https://storybook.js.org/docs/react/contribute/how-to-reproduce) doesn't seem to mention how to createe a reproduction for issue of Storybook CLI. If it is possible to create a reproduction in this case, please let me know and I am happy to create one.

### System

```shell
Environment Info:

  System:
    OS: Linux 5.10 Ubuntu 20.04 LTS (Focal Fossa)
    CPU: (8) x64 AMD Ryzen 5 3550H with Radeon Vega Mobile Gfx
  Binaries:
    Node: 16.9.1 - ~/.nvm/versions/node/v16.9.1/bin/node
    Yarn: 1.22.10 - /mnt/d/Programs/nodejs/node_global/yarn
    npm: 7.21.1 - ~/.nvm/versions/node/v16.9.1/bin/npm
```


### Additional context

_No response_

--- a/code/lib/cli/src/detect.test.ts
+++ b/code/lib/cli/src/detect.test.ts
@@ -236,17 +236,28 @@ describe('Detect', () => {
   it(`should return type HTML if html option is passed`, async () => {
     const packageManager = {
       retrievePackageJson: () => Promise.resolve({ dependencies: {}, devDependencies: {} }),
+      getAllDependencies: () => Promise.resolve({}),
       getPackageVersion: () => Promise.resolve(null),
-    } as any as JsPackageManager;
+    } as Partial<JsPackageManager>;
 
-    await expect(detect(packageManager, { html: true })).resolves.toBe(ProjectType.HTML);
+    await expect(detect(packageManager as any, { html: true })).resolves.toBe(ProjectType.HTML);
   });
 
   it(`should return language javascript if the TS dependency is present but less than minimum supported`, async () => {
     (logger.warn as jest.MockedFunction<typeof logger.warn>).mockClear();
 
     const packageManager = {
-      retrievePackageJson: () => Promise.resolve({ dependencies: {}, devDependencies: {} }),
+      retrievePackageJson: () =>
+        Promise.resolve({
+          dependencies: {},
+          devDependencies: {
+            typescript: '1.0.0',
+          },
+        }),
+      getAllDependencies: () =>
+        Promise.resolve({
+          typescript: '1.0.0',
+        }),
       getPackageVersion: (packageName) => {
         switch (packageName) {
           case 'typescript':
@@ -266,7 +277,17 @@ describe('Detect', () => {
   it(`should return language typescript-3-8 if the TS dependency is >=3.8 and <4.9`, async () => {
     await expect(
       detectLanguage({
-        retrievePackageJson: () => Promise.resolve({ dependencies: {}, devDependencies: {} }),
+        retrievePackageJson: () =>
+          Promise.resolve({
+            dependencies: {},
+            devDependencies: {
+              typescript: '3.8.0',
+            },
+          }),
+        getAllDependencies: () =>
+          Promise.resolve({
+            typescript: '3.8.0',
+          }),
         getPackageVersion: (packageName: string) => {
           switch (packageName) {
             case 'typescript':
@@ -280,7 +301,17 @@ describe('Detect', () => {
 
     await expect(
       detectLanguage({
-        retrievePackageJson: () => Promise.resolve({ dependencies: {}, devDependencies: {} }),
+        retrievePackageJson: () =>
+          Promise.resolve({
+            dependencies: {},
+            devDependencies: {
+              typescript: '4.8.0',
+            },
+          }),
+        getAllDependencies: () =>
+          Promise.resolve({
+            typescript: '4.8.0',
+          }),
         getPackageVersion: (packageName: string) => {
           switch (packageName) {
             case 'typescript':
@@ -296,7 +327,17 @@ describe('Detect', () => {
   it(`should return language typescript-4-9 if the dependency is >TS4.9`, async () => {
     await expect(
       detectLanguage({
-        retrievePackageJson: () => Promise.resolve({ dependencies: {}, devDependencies: {} }),
+        retrievePackageJson: () =>
+          Promise.resolve({
+            dependencies: {},
+            devDependencies: {
+              typescript: '4.9.1',
+            },
+          }),
+        getAllDependencies: () =>
+          Promise.resolve({
+            typescript: '4.9.1',
+          }),
         getPackageVersion: (packageName: string) => {
           switch (packageName) {
             case 'typescript':
@@ -312,7 +353,17 @@ describe('Detect', () => {
   it(`should return language typescript if the dependency is =TS4.9`, async () => {
     await expect(
       detectLanguage({
-        retrievePackageJson: () => Promise.resolve({ dependencies: {}, devDependencies: {} }),
+        retrievePackageJson: () =>
+          Promise.resolve({
+            dependencies: {},
+            devDependencies: {
+              typescript: '4.9.0',
+            },
+          }),
+        getAllDependencies: () =>
+          Promise.resolve({
+            typescript: '4.9.0',
+          }),
         getPackageVersion: (packageName: string) => {
           switch (packageName) {
             case 'typescript':
@@ -328,7 +379,17 @@ describe('Detect', () => {
   it(`should return language typescript if the dependency is =TS4.9beta`, async () => {
     await expect(
       detectLanguage({
-        retrievePackageJson: () => Promise.resolve({ dependencies: {}, devDependencies: {} }),
+        retrievePackageJson: () =>
+          Promise.resolve({
+            dependencies: {},
+            devDependencies: {
+              typescript: '4.9.0-beta',
+            },
+          }),
+        getAllDependencies: () =>
+          Promise.resolve({
+            typescript: '4.9.0-beta',
+          }),
         getPackageVersion: (packageName: string) => {
           switch (packageName) {
             case 'typescript':
@@ -345,8 +406,26 @@ describe('Detect', () => {
     await expect(
       detectLanguage({
         retrievePackageJson: () => Promise.resolve({ dependencies: {}, devDependencies: {} }),
+        getAllDependencies: () => Promise.resolve({}),
         getPackageVersion: () => {
           return Promise.resolve(null);
+        },
+      } as Partial<JsPackageManager> as JsPackageManager)
+    ).resolves.toBe(SupportedLanguage.JAVASCRIPT);
+  });
+
+  it(`should return language Javascript even when Typescript is detected in the node_modules but not listed as a direct dependency`, async () => {
+    await expect(
+      detectLanguage({
+        retrievePackageJson: () => Promise.resolve({ dependencies: {}, devDependencies: {} }),
+        getAllDependencies: () => Promise.resolve({}),
+        getPackageVersion: (packageName) => {
+          switch (packageName) {
+            case 'typescript':
+              return Promise.resolve('4.9.0');
+            default:
+              return Promise.resolve(null);
+          }
         },
       } as Partial<JsPackageManager> as JsPackageManager)
     ).resolves.toBe(SupportedLanguage.JAVASCRIPT);

--- a/code/lib/cli/src/detect.ts
+++ b/code/lib/cli/src/detect.ts
@@ -165,6 +165,10 @@ export async function detectLanguage(packageManager: JsPackageManager) {
     return language;
   }
 
+  const isTypescriptDirectDependency = await packageManager
+    .getAllDependencies()
+    .then((deps) => Boolean(deps['typescript']));
+
   const typescriptVersion = await packageManager.getPackageVersion('typescript');
   const prettierVersion = await packageManager.getPackageVersion('prettier');
   const babelPluginTransformTypescriptVersion = await packageManager.getPackageVersion(
@@ -178,20 +182,21 @@ export async function detectLanguage(packageManager: JsPackageManager) {
     'eslint-plugin-storybook'
   );
 
-  if (
-    typescriptVersion &&
-    semver.gte(typescriptVersion, '4.9.0') &&
-    (!prettierVersion || semver.gte(prettierVersion, '2.8.0')) &&
-    (!babelPluginTransformTypescriptVersion ||
-      semver.gte(babelPluginTransformTypescriptVersion, '7.20.0')) &&
-    (!typescriptEslintParserVersion || semver.gte(typescriptEslintParserVersion, '5.44.0')) &&
-    (!eslintPluginStorybookVersion || semver.gte(eslintPluginStorybookVersion, '0.6.8'))
-  ) {
-    language = SupportedLanguage.TYPESCRIPT_4_9;
-  } else if (typescriptVersion && semver.gte(typescriptVersion, '3.8.0')) {
-    language = SupportedLanguage.TYPESCRIPT_3_8;
-  } else if (typescriptVersion && semver.lt(typescriptVersion, '3.8.0')) {
-    logger.warn('Detected TypeScript < 3.8, populating with JavaScript examples');
+  if (isTypescriptDirectDependency && typescriptVersion) {
+    if (
+      semver.gte(typescriptVersion, '4.9.0') &&
+      (!prettierVersion || semver.gte(prettierVersion, '2.8.0')) &&
+      (!babelPluginTransformTypescriptVersion ||
+        semver.gte(babelPluginTransformTypescriptVersion, '7.20.0')) &&
+      (!typescriptEslintParserVersion || semver.gte(typescriptEslintParserVersion, '5.44.0')) &&
+      (!eslintPluginStorybookVersion || semver.gte(eslintPluginStorybookVersion, '0.6.8'))
+    ) {
+      language = SupportedLanguage.TYPESCRIPT_4_9;
+    } else if (semver.gte(typescriptVersion, '3.8.0')) {
+      language = SupportedLanguage.TYPESCRIPT_3_8;
+    } else if (semver.lt(typescriptVersion, '3.8.0')) {
+      logger.warn('Detected TypeScript < 3.8, populating with JavaScript examples');
+    }
   }
 
   return language;

